src/enemy.rs

```diff
@@ -0,0 +1,279 @@
+        is_elite: Option<bool>,
+        spawn_waves: Vec<SpawnWave>,
+    }
+
+    let conf = std::fs::read_to_string("configs/enemies.json").unwrap();
+    let data = serde_json::from_str::<Vec<EnemyConfig>>(&conf).unwrap();
+
+    let mut spawn_map: HashMap<u16, Vec<EnemySpawnComponent>> = HashMap::new();
+
+    for enemy_conf in data {
+        let texture_handle: Handle<Image> = asset_server.load(enemy_conf.asset_path);
+        let mut enemy = EnemySpawnComponent {
+            name: enemy_conf.name,
+            enemy: Enemy,
+            health: Health(enemy_conf.hp),
+            damage: Damage(enemy_conf.dmg),
+            max_velocity: enemy_conf.max_velocity,
+            max_force: enemy_conf.max_force,
+            mass: enemy_conf.mass,
+            texture: texture_handle,
+            is_elite: enemy_conf.is_elite,
+            rewards: Rewards { exp: 1, items: "Orange" }, // TODO: Make them drop gems)
+            timer: Default::default(),
+        };
+
+        for waves in enemy_conf.spawn_waves {
+            for n in waves.from..waves.to + 1 {
+                if let Some(mut components) = spawn_map.get_mut(&n) {
+                    enemy.timer = Timer::new(Duration::from_secs(waves.spawn_time), TimerMode::Repeating);
+                    components.push(enemy.clone());
+                } else {
+                    let mut spawn_vec = Vec::new();
+                    enemy.timer = Timer::new(Duration::from_secs(waves.spawn_time), TimerMode::Repeating);
+                    spawn_vec.push(enemy.clone());
+                    spawn_map.insert(n, spawn_vec);
+                }
+            }
+        }
+    }
+
+    let spawners = EnemySpawners(spawn_map);
+    commands.insert_resource(spawners.clone());
+}
+fn spawn(
+    mut commands: Commands,
+    mut spawn_map: ResMut<EnemySpawners>,
+    mut player: Query<&mut Transform, With<Player>>,
+    mut current_wave: ResMut<CurrentWave>,
+    mut global_time_ticker: Res<GlobalTimeTickerResource>,
+) {
+    if global_time_ticker.0.finished() || current_wave.need_wave_spawn {
+        if let Some(spawners) = spawn_map.0.get_mut(&current_wave.num) {
+            for mut spawner in spawners {
+                spawner.timer.tick(global_time_ticker.0.duration());
+
+                if spawner.timer.finished() || current_wave.need_wave_spawn == true {
+                    // spawn elite only one time on the wave
+                    if Some(true) == spawner.is_elite && current_wave.need_wave_spawn == false {
+                        continue
+                    }
+                    if let Ok(p_t) = player.get_single_mut() {
+                        let is_left = thread_rng().gen_range(0, 2);
+                        let is_up = thread_rng().gen_range(0, 2);
+                        let mut enemy_batch: Vec<(EnemyBundle, SpriteBundle, SteeringBundle, Name)> = Vec::new();
+
+                        for i in 1..thread_rng().gen_range(ENEMY_BATCH_SIZE.min, ENEMY_BATCH_SIZE.max) {
```

Comment:
You can't use `f32` to specify rng range. Cast to `i32`.

Reply:
changed to generic type